### PR TITLE
Add Default Service Sinks and Sources

### PR DESCRIPTION
**Pull request by niruhan**
**Sinks**
- grpc: fire and forget type sink
- grpc-call: send request to a grpc server and get a response back
- grpc-service-response: send responses back for requests received via grpc-service source

**Sources**
- grpc: receive: request and send downstream
- grpc-service: receive request and send responses back
- grpc-call-response: listens to responses for requests sent through grpc-call sink

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/pom.xml [hunk 1]
```diff
@@ -0,0 +1,173 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dnwick) [inline component/pom.xml]: need to have license

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
+package io.siddhi.extension.io.grpc.sink;
```
Comment (dnwick) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: License please

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java [hunk 3]
```diff
@@ -148,14 +150,23 @@ public void process(Event request,
...
+            this.receivedTime = receivedTime;
+        }
+
         @Override
         public void run() {
-            for (String messageId: timestampOfMessageIds.keySet()) {
-                if (timestampOfMessageIds.get(messageId) < siddhiAppContext.getTimestampGenerator().currentTime() -
-                        serviceTimeout) {
-                    streamObserverMap.get(messageId).onError(new io.grpc.StatusRuntimeException(
+            if (receivedTime < siddhiAppContext.getTimestampGenerator().currentTime() - serviceTimeout) {
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: We need to wait in sleep while loop till this condition is met (if the timer has released the value earlier)

Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: I am starting this thread with an initial delay of service.timeout. So anyway this condition will be checked after service.timeout

Comment (suhothayan) [reply component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: So why we need to do the if check?  
siddhiAppContext.getTimestampGenerator().currentTime() can be varied based on the playback so we can always assume that. Implement a while loop to handle this.

File: README.md [hunk 4]
```diff
@@ -1 +1,44 @@
-# siddhi-io-grpc
\ No newline at end of file
+﻿siddhi-io-grpc
+======================================
+
+
+
```
Comment (suhothayan) [inline README.md]: Update the readme based on the new theme (see https://github.com/siddhi-io/siddhi-io-http)

File: component/pom.xml [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.extension.siddhi.io.grpc.*,
```
Comment (suhothayan) [inline component/pom.xml]: Fix the export packages

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    protected StateFactory init(StreamDefinition streamDefinition, OptionHolder optionHolder, ConfigReader configReader,
+                                SiddhiAppContext siddhiAppContext) {
+        this.siddhiAppContext = siddhiAppContext;
+        this.url = optionHolder.validateAndGetOption(GrpcConstants.PUBLISHER_URL).getValue();
+        this.streamID = siddhiAppContext.getName() + GrpcConstants.PORT_HOST_SEPARATOR + streamDefinition.getId();
+        if (optionHolder.isOptionExists(GrpcConstants.HEADERS)) {
+            this.headersOption = optionHolder.validateAndGetOption(GrpcConstants.HEADERS);
+        }
+        if (!url.substring(0, 4).equalsIgnoreCase(GrpcConstants.GRPC_PROTOCOL_NAME)) {
+            throw new SiddhiAppValidationException(streamID + "The url must begin with \"" +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: Add Siddhi App name and streamID for all the exceptions, also space is needed here `streamID + "The url`. 
Formate the text correctly !!!!

Comment (suhothayan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: Check all the exceptions and add Siddhi App name and streamID and formate the text correctly !!!!

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+import org.wso2.grpc.EventServiceGrpc.EventServiceFutureStub;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@code GrpcCallSink} Handle the gRPC publishing tasks and injects response into grpc-call-response source.
+ */
+@Extension(
+        name = "grpc-call", namespace = "sink",
+        description = "This extension publishes event data encoded into GRPC Classes as defined in the user input " +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: Can you simply this. and than explaining can you give the proto file here

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                "Event and receives a response of the same type. Consume sends a request of type Event and expects " +
+                "no response from gRPC server. Please note that the Event type mentioned here is not " +
+                "io.siddhi.core.event.Event but a type defined in the default service protobuf given in the readme. " +
+                "This grpc-call sink is used for scenarios where we send a request out and expect a response back. " +
+                "In default mode this will use EventService process method.",
+        parameters = {
+                @Parameter(name = "url",
+                        description = "The url to which the outgoing events should be published via this extension. " +
+                                "This url should consist the host address, port, service name, method name in the " +
+                                "following format. grpc://hostAddress:port/serviceName/methodName" ,
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: Use the same format used in HTTP io

Comment (suhothayan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: Check all fix all the docs

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                        description = "Sets the maximum size of metadata allowed to be received in bytes" ,
+                        type = {DataType.LONG},
+                        optional = true,
+                        defaultValue = "8192"),
+        },
+        examples = {
+                @Example(
+                        syntax = "@sink(type='grpc-call', " +
+                                "url = 'grpc://194.23.98.100:8080/EventService/process', " +
+                                "sink.id= '1', @map(type='json')) "
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: sink.id not explained in the param list !!! 

Have the example with grpc-call and grpc-call-response

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        }
+    }
+
+    @Override
+    public void publish(Object payload, DynamicOptions dynamicOptions, State state)
+            throws ConnectionUnavailableException {
+        if (isDefaultMode) {
+            if (!(payload instanceof String)) {
+                throw new SiddhiAppRuntimeException(siddhiAppContext.getName() + ": Payload should be of type String " +
+                        "for default EventService but found " + payload.getClass().getName());
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: Improve logs with stream id

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+            Futures.addCallback(futureResponse, new FutureCallback<Event>() {
+                @Override
+                public void onSuccess(Event result) {
+                    GrpcSourceRegistry.getInstance().getGrpcCallResponseSourceSource(sinkID).onResponse(result);
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug(siddhiAppContext.getName() + ": " + t.getMessage());
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: Stream id

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    public abstract void initSink(OptionHolder optionHolder);
+
+    /**
+     * This method will be called before the processing method.
+     * Intention to establish connection to publish event.
+     * @throws ConnectionUnavailableException if end point is unavailable the ConnectionUnavailableException thrown
+     *                                        such that the  system will take care retrying for connection
+     */
+    @Override
+    public void connect() throws ConnectionUnavailableException {}
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: Remove this here

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+    /**
+     * Called after all publishing is done, or when {@link ConnectionUnavailableException} is thrown
+     * Implementation of this method should contain the steps needed to disconnect from the sink.
+     */
+    @Override
+    public void disconnect() {}
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: Remove this here as you are overriding them

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+     * Intention to establish connection to publish event.
+     * @throws ConnectionUnavailableException if end point is unavailable the ConnectionUnavailableException thrown
+     *                                        such that the  system will take care retrying for connection
+     */
+    @Override
+    public void connect() throws ConnectionUnavailableException {
+        this.channel = ManagedChannelBuilder.forTarget(address).usePlaintext().build();
+        this.futureStub = EventServiceGrpc.newFutureStub(channel);
+        if (!channel.isShutdown()) {
+            logger.info(streamID + " has successfully connected to " + url);
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: streamID is not conected the` grpc... on streamID` is

Comment (suhothayan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: if channel.isShutdown() then what to do?? dont we need  an error ?

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+ */
+@Extension(
+        name = "grpc-service-response", namespace = "sink",
+        description = "This extension is used to send responses back to a gRPC client after receiving requests " +
+                "through grpc-service source. This calls a callback in grpc-service source to put response back in " +
+                "StreamObserver.",
+        parameters = {
+                @Parameter(name = "url",
+                        description = "The url to which the outgoing events should be published via this extension. " +
+                                "This url should consist the host address, port, service name, method name in the " +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java]: Use the same form of doc used in HTTP IO

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.log4j.Logger;
+
+/**
+ * {@code GrpcServiceResponseSink} Handle sending responses for requests received via grpc-service source.
+ */
+@Extension(
+        name = "grpc-service-response", namespace = "sink",
+        description = "This extension is used to send responses back to a gRPC client after receiving requests " +
+                "through grpc-service source. This calls a callback in grpc-service source to put response back in " +
+                "StreamObserver.",
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java]: You dont need to talk about internals sush as`StreamObserver `

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    @Override
+    public void publish(Object payload, DynamicOptions dynamicOptions, State state) {
+        String messageId = messageIdOption.getValue(dynamicOptions);
+        grpcSourceRegistry.getGrpcServiceSource(sourceId).handleCallback(messageId, (String) payload);
+    }
+
+    @Override
+    public void connect() throws ConnectionUnavailableException {
+    }
+    @Override
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java]: space between methods !!!!

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+                        type = {DataType.STRING}),
+                @Parameter(name = "source.id",
+                        description = "A unique id to identify the correct source to which this sink is mapped. " +
+                                "There is a 1:1 mapping between source and sink" ,
+                        type = {DataType.INT}),
+        },
+        examples = {
+                @Example(
+                        syntax = "@sink(type='grpc-service-response', " +
+                                "url = 'grpc://134.23.43.35:8080/org.wso2.grpc.EventService/consume/mySequence', " +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcServiceResponseSink.java]: have a grpc-service also in the example else people will not understand

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+ * {@code GrpcSink} Handle the gRPC publishing tasks.
+ */
+@Extension(
+        name = "grpc", namespace = "sink",
+        description = "This extension publishes event data encoded into GRPC Classes as defined in the user input " +
+                "jar. This extension has a default gRPC service classes added. The default service is called " +
+                "\"EventService\" and it has 2 rpc's. They are process and consume. Process sends a request of type " +
+                "Event and receives a response of the same type. Consume sends a request of type Event and expects " +
+                "no response from gRPC server. Please note that the Event type mentioned here is not " +
+                "io.siddhi.core.event.Event but org.wso2.grpc.Event. This grpc sink is used for scenarios where we " +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: do the same fixes mentioned for ther docs

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+                        description = "A similar example to above but with headers. Headers are also send into the " +
+                                "stream as a data. In the sink headers dynamic property reads the value and sends " +
+                                "it as MetaData with the request"
+                )
+        }
+)
+
+public class GrpcSink extends AbstractGrpcSink {
+    private static final Logger logger = Logger.getLogger(GrpcSink.class.getName());
+    EventServiceStub asyncStub;
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: Why not private?

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+     * Intention to establish connection to publish event.
+     * @throws ConnectionUnavailableException if end point is unavailable the ConnectionUnavailableException thrown
+     *                                        such that the  system will take care retrying for connection
+     */
+    @Override
+    public void connect() throws ConnectionUnavailableException {
+        this.channel = managedChannelBuilder.build();
+        this.asyncStub = EventServiceGrpc.newStub(channel);
+        if (!channel.isShutdown()) {
+            logger.info(streamID + " has successfully connected to " + url);
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcSink.java]: Fix log, check other logs too

File: component/src/main/java/io/siddhi/extension/io/grpc/source/AbstractGrpcSource.java [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+    private String serviceName;
+    protected boolean isDefaultMode;
+    private int port;
+    protected SourceServerInterceptor serverInterceptor;
+    protected ServerBuilder serverBuilder;
+    protected int serverShutdownWaitingTime;
+
+    @Override
+    protected ServiceDeploymentInfo exposeServiceDeploymentInfo() {
+        return null;
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/AbstractGrpcSource.java]: This should not be null it should expose the service info (need to add for other sources too)

File: component/src/main/java/io/siddhi/extension/io/grpc/source/AbstractGrpcSource.java [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+    public abstract void populateHeaderString(String headerString);
+
+    /**
+     * Returns the list of classes which this source can output.
+     *
+     * @return Array of classes that will be output by the source.
+     * Null or empty array if it can produce any type of class.
+     */
+    @Override
+    public Class[] getOutputEventClasses() {
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/AbstractGrpcSource.java]: If methods are overridden in child classes remove them here

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcCallResponseSource.java [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+                        description = "a unique ID that should be set for each gRPC source. There is a 1:1 mapping " +
+                                "between gRPC sinks and sources. Each sink has one particular source listening to " +
+                                "the responses to requests published from that sink. So the same sink.id should be " +
+                                "given when writing the sink also." ,
+                        type = {DataType.INT}),
+        },
+        examples = {
+                @Example(
+                        syntax = "@source(type='grpc-call-response', sink.id= '1') " +
+                                "define stream BarStream (message String);",
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcCallResponseSource.java]: Fix docs as in HTTP IO

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+        this.timer = new Timer();
+        GrpcSourceRegistry.getInstance().putGrpcServiceSource(sourceId, this);
+    }
+
+    @Override
+    public void connect(ConnectionCallback connectionCallback, State state) throws ConnectionUnavailableException {
+        try {
+            server.start();
+            if (logger.isDebugEnabled()) {
+                logger.debug(siddhiAppContext.getName() + ": Server started");
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: fix logs with stream name

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+                                "@map(type='json', @attributes(name='trp:name', age='trp:age', message='message'))) " +
+                                "define stream BarStream (message String, name String, age int);",
+                        description = "Here we are getting headers sent with the request as transport properties and " +
+                                "injecting them into the stream. With each request a header will be sent in " +
+                                "MetaData in the following format: 'Name:John', 'Age:23'"
+                )
+        }
+)
+public class GrpcServiceSource extends AbstractGrpcSource {
+    private static final Logger logger = Logger.getLogger(GrpcServiceSource.class.getName());
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: service info should be exposed

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    @Override
+    public void populateHeaderString(String headerString) {
+        this.headerString = headerString;
+    }
+
+    public void handleCallback(String messageId, String responsePayload) {
+        if (isDefaultMode) {
+            readWriteLock.writeLock().lock();
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: Why we need to have a readWriteLock?

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+            try {
+                StreamObserver<Event> streamObserver = streamObserverMap.get(messageId);
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: Cant we do a sync map and only do remove here that will return the removed StreamObserver

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcSource.java [hunk 29]
```diff
@@ -0,0 +1,179 @@
...
+@Extension(
+        name = "grpc",
+        namespace = "source",
+        description = "This extension starts a grpc server during initialization time. The server listens to " +
+                "requests from grpc stubs. This source has a default mode of operation and custom user defined grpc " +
+                "service mode. In the default mode this source will use EventService consume method. This method " +
+                "will receive requests and injects them into stream through a mapper.",
+        parameters = {
+                @Parameter(name = "url",
+                        description = "The url which can be used by a client to access the grpc server in this " +
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcSource.java]: Fix docs, logs, and exceptions

File: component/src/main/java/io/siddhi/extension/io/grpc/util/GrpcSourceRegistry.java [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.HashMap;
+
+/**
+ * a class to register GrpcSource with respective sink.id or source.id. Used by GrpcCallSink and GrpcServiceResponseSink
+ * to push responses
+ */
+public class GrpcSourceRegistry {
+    private static GrpcSourceRegistry instance = new GrpcSourceRegistry();
+    private HashMap<String, GrpcCallResponseSource> grpcCallResponseSourceHashMap = new HashMap<>();
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/util/GrpcSourceRegistry.java]: Use sync maps

File: component/src/main/java/io/siddhi/extension/io/grpc/util/GrpcUtils.java [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * Class to hold the static util methods needed
+ */
+public class GrpcUtils {
+    public static String getServiceName(String path) {
+        List<String> urlParts = new ArrayList<>(Arrays.asList(path.split(GrpcConstants.PORT_SERVICE_SEPARATOR)));
+        urlParts.removeAll(Collections.singletonList(GrpcConstants.EMPTY_STRING));
+        if (urlParts.size() < 2) {
+            throw new SiddhiAppValidationException("Malformed URL. After port number atleast two sections should " +
+                    "be available seperated by / as follows grpc://host:port/ServiceName/MethodName");
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/util/GrpcUtils.java]: Say `by '/' as in 'grpc://<host>:<port>/<ServiceName>/<MethodName>'"` fix other placess too

File: component/src/main/java/io/siddhi/extension/io/grpc/util/SourceServerInterceptor.java [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> serverCall,
+                                                               Metadata metadata,
+                                                               ServerCallHandler<ReqT, RespT> serverCallHandler) {
+    Metadata.Key<String> headerKey = Metadata.Key.of(GrpcConstants.HEADERS, ASCII_STRING_MARSHALLER);
+    associatedGrpcSource.populateHeaderString(metadata.get(headerKey));
+    if (logger.isDebugEnabled()) {
+      logger.debug("Header received: " + metadata.get(headerKey));
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/util/SourceServerInterceptor.java]: Fix all the logs !!!

File: component/src/main/java/org/wso2/grpc/EventServiceGrpc.java [hunk 33]
```diff
@@ -0,0 +1,355 @@
+package org.wso2.grpc;
+
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/grpc/EventServiceGrpc.java]: If its GRPC generated class add a comment and also all files need to have license !!

File: component/src/main/java/org/wso2/grpc/EventServiceOuterClass.java [hunk 34]
```diff
@@ -0,0 +1,62 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
+// source: EventService.proto
+
+package org.wso2.grpc;
```
Comment (suhothayan) [inline component/src/main/java/org/wso2/grpc/EventServiceOuterClass.java]: License

File: mkdocs.yml [hunk 35]
```diff
@@ -0,0 +1,30 @@
+site_name: Siddhi IO Grpc
+site_description: Siddhi IO Grpc Extension
```
Comment (suhothayan) [inline mkdocs.yml]: Use the correct way to write Grpc!!! in the site

Comment (suhothayan) [reply mkdocs.yml]: User the same theme as in HTTP IO

File: pom.xml [hunk 36]
```diff
@@ -0,0 +1,208 @@
...
+            <dependency>
+                <groupId>io.siddhi.extension.map.json</groupId>
+                <artifactId>siddhi-map-json</artifactId>
+                <version>${siddhi.map.json.version}</version>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>6.11</version>
```
Comment (suhothayan) [inline pom.xml]: Version should be in properties

File: pom.xml [hunk 37]
```diff
@@ -0,0 +1,208 @@
...
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+    <properties>
+        <grpc.version>1.22.1</grpc.version>
+        <transport.http.version>6.0.294</transport.http.version>
+        <mavan.findbugsplugin.exclude.file>findbugs-exclude.xml</mavan.findbugsplugin.exclude.file>
+        <siddhi.version>5.0.0</siddhi.version>
```
Comment (suhothayan) [inline pom.xml]: Use the latest siddhi version

File: component/src/test/java/io/siddhi/extension/io/grpc/sink/GrpcSinkTestCase.java [hunk 38]
```diff
@@ -0,0 +1,298 @@
...
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class GrpcSinkTestCase {
+    private static final Logger log = Logger.getLogger(GrpcSinkTestCase.class.getName());
+    private TestServer server = new TestServer();
```
Comment (suhothayan) [inline component/src/test/java/io/siddhi/extension/io/grpc/sink/GrpcSinkTestCase.java]: Make sure to add all test cases to the build

File: component/src/test/java/io/siddhi/extension/io/grpc/utils/TestServer.java [hunk 39]
```diff
@@ -0,0 +1,82 @@
+package io.siddhi.extension.io.grpc.utils;
+
```
Comment (suhothayan) [inline component/src/test/java/io/siddhi/extension/io/grpc/utils/TestServer.java]: License is missing. 
Damth has pointed out the license issue, and its bad you have not fixed that!

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java (hunk 3 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: we don't need the if check. After the service timeout the existence of that stream observer will be enough to decide to send error. Will remove if

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java (hunk 6 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/AbstractGrpcSink.java]: I have added siddhi app name in the streamID. will remove the siddhi app name part and add seperately in each error

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java (hunk 14 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: I moved the channel building part into connect(). Previously did this in init() thats why I added the if check. Will remove the if check but keep the modified log

File: component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java (hunk 27 above)
Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: the streamObserverMap will be accessed by main thread in this handleCallback as well as timer task

Comment (niruhan) [reply component/src/main/java/io/siddhi/extension/io/grpc/source/GrpcServiceSource.java]: ignore the previous comment. Implemented sync map and remove based retrieval

File: component/src/main/java/org/wso2/grpc/EventServiceGrpc.java (hunk 33 above)
Comment (niruhan) [reply component/src/main/java/org/wso2/grpc/EventServiceGrpc.java]: Is it okay to add wso2 licenses for autogenerated classes?

File: component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java [hunk 40]
```diff
@@ -46,113 +46,157 @@
...
                         defaultValue = "4194304"),
-                @Parameter(name = "max.inbound.metadata.size",
+                @Parameter(
+                        name = "max.inbound.metadata.size",
                         description = "Sets the maximum size of metadata allowed to be received in bytes" ,
                         type = {DataType.LONG},
                         optional = true,
                         defaultValue = "8192"),
+                @Parameter(
+                        name = "sink.id",
```
Comment (suhothayan) [inline component/src/main/java/io/siddhi/extension/io/grpc/sink/GrpcCallSink.java]: move this up

File: component/src/main/java/org/wso2/grpc/EventServiceGrpc.java (hunk 33 above)
Comment (suhothayan) [reply component/src/main/java/org/wso2/grpc/EventServiceGrpc.java]: yes

**Pull request closed without merging**